### PR TITLE
perf(stock): skip the negative batch scan when nothing is negative

### PR DESCRIPTION
**Pull request by mihir-kandoi**
Every inward row copied the whole FIFO queue and scanned it for a negative slot of that batch, so replaying a warehouse cost time quadratic in the number of slots held there. Almost no ledger has a negative batch slot, so almost all of that work found nothing.

Profile of 4000 inward entries across 4000 batches for one item and warehouse:

```
ncalls    tottime  cumtime
7998000    1.604    3.953   _is_matching_negative_batch_slot
   4000    0.674    4.630   _neutralize_negative_batch_stock
```

`_append_negative_batch_slot` is the only place a batch slot goes negative, so it records the batch and warehouse, and the scan runs only for a batch recorded there. The record is discarded again by a walk that reaches the end of the queue with nothing negative left behind; a walk that stops early has not seen the rest of the queue and keeps it. So a batch can be marked only by the code that makes it negative, and unmarked only by a walk that proved it is not.

4000 receipts of batches that never went negative:

```
before  0.92s   ~4x per doubling
after   0.01s   1.1x, 1.7x per doubling
```

4000 receipts of one batch after a single brief negative, which is what the record has to stop gating:

```
without the discard  0.77s   4.0x per doubling
with the discard     0.01s   2.1x per doubling
```

No behaviour change. The existing negative stock tests cover the gated path, plus one new test for a negative slot left behind by a partial refill.

### CODE REVIEW

File: erpnext/stock/report/stock_ageing/stock_ageing.py [hunk 1]
```diff
@@ -791,9 +795,13 @@ def _append_negative_batch_slot(
 		qty: float,
 		stock_value_difference: float,
 	) -> None:
+		"""The only place a batch slot goes negative, so it is also where the warehouse
+		is recorded as owing stock on that batch. A key that outlives the negative slot
+		only costs the scan that ran before."""
 		fifo_queue.append(
 			[batch_no, use_batchwise_valuation, -(qty), row.posting_date, -(stock_value_difference)]
 		)
+		self.batches_with_negative_slots.add((batch_no, row.warehouse))
```
[the remark this answers is not in the document]
Comment (mihir-kandoi) [reply erpnext/stock/report/stock_ageing/stock_ageing.py:812]: Good catch, this was real. The record is now discarded when a walk reaches the end of the queue with no negative slot of that batch left behind; a walk that stops early keeps it, since it has not seen the rest of the queue.

Measured with 4000 receipts of one batch after a single brief negative: 0.77s and still 4.0x per doubling before, 0.01s and linear after. Added a test for the case the conservative half protects, where a partial refill leaves a second negative slot for a later receipt to find.
